### PR TITLE
musicutil.lua and scales demo script

### PR DESCRIPTION
**Pull request by markwheeler**
Put musicutil.lua and associated demo script together into dust. Only small tweaks since yesterday.

Also generated a quick ldoc page for musicutil - not sure if this is a good idea or confusing given that the other docs in that folder are MD and for scripts and engines. Included for discussion 🙂

### CODE REVIEW

File: lib/lua/mark_eats/musicutil.lua [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    output[i] = name
+  end
+  if note_nums_len == 1 then return output[1]
+  else return output end
+end
+
+--- Convert MIDI note numbers to frequencies.
+-- @param note_nums MIDI note number (0-127) or array of note numbers.
+-- @return Frequency number in Hz or array of frequencies.
+function MusicUtil.note_nums_to_freqs(note_nums)
```
Comment (simonvanderveldt) [inline lib/lua/mark_eats/musicutil.lua]: Was wondering if there's any particular reason this takes multiple notes as input? Can't the script just call the function as many times as it needs it?
And if there is merit to having it as a library shouldn't it just be a separate function that calls the simple/one note input function multiple times?

Comment (markwheeler) [reply lib/lua/mark_eats/musicutil.lua]: It was just for convenience, but you're right, it's unusual to have methods returning varying types like that, I got a little over enthusiastic with the dynamic typing there.  😄 
I just pushed a commit that refactors these into 'single item' methods and 'array methods' for clarity. Thx for the feedback!

File: lib/lua/mark_eats/musicutil.lua [hunk 2]
```diff
@@ -107,91 +107,110 @@ function MusicUtil.generate_scale(root_num, scale_type, octaves)
...
-          note_nums[n] = note_nums[n] + prev_delta
-          break
-        end
-        prev_delta = delta
-      end
-    end
-  end
-  if note_nums_len == 1 then return note_nums[1]
-  else return note_nums end
+	if snap_array_len == 1 then
```
Comment (pq) [inline lib/lua/mark_eats/musicutil.lua]: nit: is this indent intended?

File: lib/lua/mark_eats/musicutil.lua [hunk 3]
```diff
@@ -107,91 +107,110 @@ function MusicUtil.generate_scale(root_num, scale_type, octaves)
...
+
+  return note_num
+end
+
+--- Snap an array of MIDI note numbers to an array of note numbers.
+-- @param note_nums_array Array of input MIDI note numbers.
+-- @param snap_array Array of MIDI note numbers to snap to, must be in low to high order.
+-- @return Array of adjusted note numbers.
+function MusicUtil.snap_notes_to_array(note_nums_array, snap_array)
+	for i = 1, #note_nums_array do
```
Comment (pq) [inline lib/lua/mark_eats/musicutil.lua]: similarly, this looks mega-indented.

(we should really iron-out a formatter solution)

are you using a maiden or working locally?

File: scripts/mark_eats/scales_demo.lua [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  if step > scale_len then
+    step = scale_len
+    step_increment = -1
+  else
+    step = step + step_increment
+    if step == 1 or step == scale_len then
+      step_increment = step_increment * -1
+    end
+  end
+  engine.hz(MusicUtil.note_num_to_freq(scale_notes[step]))
```
Comment (pq) [inline scripts/mark_eats/scales_demo.lua]: aside: it'd be interesting to consider adding a `note` function to "PolyPerc" (and related engines).  i bet lots of folks would find that handy!

File: lib/lua/mark_eats/musicutil.lua (hunk 3 above)
Comment (markwheeler) [reply lib/lua/mark_eats/musicutil.lua]: ah yes sorry i messed it up in sublime

File: scripts/mark_eats/scales_demo.lua (hunk 4 above)
Comment (markwheeler) [reply scripts/mark_eats/scales_demo.lua]: Yes I think having proper consistency between the engines would be a very good idea. At least making everything respond to `freq` and `gate` which seems standard in SC.

File: lib/lua/mark_eats/musicutil.lua (hunk 3 above)
Comment (pq) [reply lib/lua/mark_eats/musicutil.lua]: out of curiosity, do you use a prettifier / formatter in sublime?

Comment (markwheeler) [reply lib/lua/mark_eats/musicutil.lua]: I didn't research it much, just installed this package but looking at that link it seems pretty unmaintained! https://github.com/rorydriscoll/LuaSublime

**Pull request merged**